catalog: enable leased descriptors by default

fqazi opened the pull request:
Previously, the catalog views inside CRDB would use KV batch operations to read descriptors for crdb_internal / pg_catalog / information_schema views. This limited how many descriptors these functions could support, since there would be no caching across transactions. This patch enables this setting by leased descriptors and locked descriptor leasing by default. This patch contains the following changes:

1. Fixes for resolving names for recently created objects when locked descriptor leasing is enabled.
2. Support for releasing locked descriptor timestamps for certain operations at the cost of correctness
3. Optional support for disabling retention of old versions of descriptors, which is there in case (2) becomes a problem. This would instead lead to retryable errors when descriptors needed for a transaction are modified before you can lease them.
4. Logic to generate retry errors if the leased descriptors and KV versions don't match for schema changes, which allows transactions to work correctly when using leased descriptors.
5. Enabling locked descriptor leasing and leased descriptors by default and updating benchmark expectations for them.

Release note (sql change): The default value of
`sql.catalog.allow_leased_descriptors.enabled` is now true. This setting
allows introspection tables like information_schema and pg_catalog to
use cached descriptors when building the table results, which improves
the performance of introspection queries when there are many tables in
the cluster.

Fixes: #159161

fqazi committed:
catalog/lease: fix name resolution bug with locked leasing

Previously, we would only attempt to look up names using the leased
timestamp when AcquireByName was invoked. This worked fine for existing
objects, but recently created objects could incorrectly fail on lookups.
To address this, this patch adds extra logic to retry the base timestamp
if the object does not exist at the leased timestamp. This mimics
behavior of ID based lookups

Release note: None

fqazi committed:
catalog: enable leased descriptors by default

Previously, leased descriptors were only enabled in a subset of tests.
This patch enables them by default in CRDB.

Fixes: #159161
Release note (sql change): The default value of
`sql.catalog.allow_leased_descriptors.enabled` is now true. This setting
allows introspection tables like information_schema and pg_catalog to
use cached descriptors when building the table results, which improves
the performance of introspection queries when there are many tables in
the cluster.

fqazi committed:
catalog/lease: add support for releasing timestamps early

Previously, crdb_internal.reset_sql_stats and
crdb_internal.reset_activity_tables could take a long time to execute in
locked leasing mode. This occurred because they initiated a schema
change in a separate transaction and waited for its completion. In
locked leasing mode, the timestamps remained in use, as we waited for
leases to expire.  To address this, this patch allows these methods to
release the timestamps earlier.

Release note: None

fqazi committed:
catalog/lease: add support for disabling old version retention

Previously, the locked leasing mode would hold old versions, until there
was no possibility of old transactions access prior descriptor versions.
This could be problematic if there are long running transactions, since
we would end up waiting for leases to expire more frequently. This patch
adds the ability to avoid holding prior versions, which as a trade off
would cause in flight transactions to hit retries, if they can't get a
valid descriptor. This is old version retention is disabled by default,
since it behaves better, in exchange for the possibility of transactions
impacted by schema changes retrying.

Release note: None

rafiss reviewed (changes requested):
nice work!

all my comments are very minor

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/159162#-:-OgEJK0M-cB7WU1jBhn5:b5qj80w)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @fqazi)

___
*[`-- commits` line 44 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/159162#-OgEIhqL68LUS0gQPQMG:-OgEIhqL68LUS0gQPQMH:b-qlqau6):*
nit: we need a `Release note (sql change)` for this commit. something like:

```
Release note (sql change): The default value of `sql.catalog.allow_leased_descriptors.enabled` is now
true. This setting allows introspection tables like information_schema and pg_catalog to use cached
descriptors when building the table results, which improves the performance of introspection queries
when there are many tables in the cluster.
```
___
*[`pkg/sql/repair.go` line 635 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/159162#-OgEHkozAoEBKdqSpimD:-OgEHkozAoEBKdqSpimE:b-rsbcgx) ([raw file](https://github.com/cockroachdb/cockroach/blob/f3086f2fadf5b0e00f9fb1aaae744266bf86e1d9/pkg/sql/repair.go#L635)):*
> ```Go
> }
> 
> // ResetLeaseTimestamp is part of interface Planner.
> ```

nit: i think 
___
*[`pkg/sql/catalog/descs/leased_descriptors.go` line 85 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/159162#-OgEI2f85mLqoldr-JZs:-OgEI2f85mLqoldr-JZt:b-rhuq1w) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c8639c230c6d2b4b1736450b5ad79915a2ce903/pkg/sql/catalog/descs/leased_descriptors.go#L85)):*
> ```Go
> 
> // leasedDescriptors holds references to all the descriptors leased in the
> //
> ```

nit: extra space
___
*[`pkg/sql/catalog/descs/leased_descriptors.go` line 109 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/159162#-OgEIE1c5XP2d8OpOfvl:-OgEIE1c5XP2d8OpOfvm:b-20pb3y) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c8639c230c6d2b4b1736450b5ad79915a2ce903/pkg/sql/catalog/descs/leased_descriptors.go#L109)):*
> ```Go
> // Error implements error interface.
> func (r *retryOnModifiedDescriptor) Error() string {
> 	return fmt.Sprint(errors.Formattable(r))
> ```

nit: the canonical approach here is: `return redact.Sprint(e).StripMarkers()`
___
*[`pkg/sql/catalog/descs/leased_descriptors.go` line 114 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/159162#-OgEIHwD1gDrYwX4B3fT:-OgEIHwD1gDrYwX4B3fU:bjkgjgd) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c8639c230c6d2b4b1736450b5ad79915a2ce903/pkg/sql/catalog/descs/leased_descriptors.go#L114)):*
> ```Go
> // SafeFormatError implements SafeFormatter.
> func (r *retryOnModifiedDescriptor) SafeFormatError(p errors.Printer) (next error) {
> 	p.Printf("the descriptor %s(%d) has been modified at timestamp %s, which is no longer usable by the txns timestamp: %s",
> ```

nit: change to "no longer usable by the transaction's timestamp"
___
*[`pkg/sql/catalog/descs/leased_descriptors.go` line 163 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/159162#-OgEIQNNAhTj0ZWRHP4E:-OgEIQNNAhTj0ZWRHP4F:b-20pb3y) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c8639c230c6d2b4b1736450b5ad79915a2ce903/pkg/sql/catalog/descs/leased_descriptors.go#L163)):*
> ```Go
> // Error implements error interface.
> func (e *mismatchedExternalDataRowTimestamp) Error() string {
> 	return fmt.Sprint(errors.Formattable(e))
> ```

nit: the canonical approach here is: `return redact.Sprint(e).StripMarkers()`
___
*[`pkg/sql/sem/eval/deps.go` line 290 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/159162#-OgEHrLfBSeVQBXGt2Mg:-OgEHrLfBSeVQBXGt2Mh:bq7vinl) ([raw file](https://github.com/cockroachdb/cockroach/blob/f3086f2fadf5b0e00f9fb1aaae744266bf86e1d9/pkg/sql/sem/eval/deps.go#L290)):*
> ```Go
> 
> 	// ResetLeaseTimestamp is used to release the locked lease timestamp.
> 	// Note: This should only be use if we have a deadlock between the current txn
> ```

nit: "be used" instead of "be use"
___
*[`pkg/sql/sem/eval/deps.go` line 292 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/159162#-OgEGwYP6mSPeCsZE9za:-OgEGwYP6mSPeCsZE9zb:b96fvb5) ([raw file](https://github.com/cockroachdb/cockroach/blob/f3086f2fadf5b0e00f9fb1aaae744266bf86e1d9/pkg/sql/sem/eval/deps.go#L292)):*
> ```Go
> 	// Note: This should only be use if we have a deadlock between the current txn
> 	// and another txn conducting a schema change on our behalf. The correct way
> 	// to achieve this is to execute the schema change.
> ```

just to clarify, what is "achieve this" referring to?
___
*[`pkg/sql/sem/builtins/pg_builtins.go` line 1432 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/159162#-OgEH1O87IZBHp29jRar:-OgEH1O87IZBHp29jRas:bzdiv15) ([raw file](https://github.com/cockroachdb/cockroach/blob/f3086f2fadf5b0e00f9fb1aaae744266bf86e1d9/pkg/sql/sem/builtins/pg_builtins.go#L1432)):*
> ```Go
> 				// Release the locked timestamp before sleeping, so that we aren't
> 				// holding up any schema changes:
> 				// (TODO)fqazi: This can mean we can mixed descriptor versions after this
> ```

what's the impact of this TODO? meaning, if you try to do introspection after the pg_sleep, it might have weird results?
___
*[`pkg/sql/catalog/descs/leased_descriptors.go` line 202 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/159162#-OgEHSl7612D_HIj4ypt:-OgEHSl7612D_HIj4ypu:b33t18d) ([raw file](https://github.com/cockroachdb/cockroach/blob/f3086f2fadf5b0e00f9fb1aaae744266bf86e1d9/pkg/sql/catalog/descs/leased_descriptors.go#L202)):*
> ```Go
> 	} else if ld.leaseTimestampSet {
> 		// Timestamp is already set so release it before picking a new one.
> 		// TODO (fqazi): Add handling for moving timestamps forward.
> ```

nit: should we get a tracking issue for moving timestamps forward? it feels a bit more complex than just a TODO.
___
*[`pkg/sql/sem/builtins/builtins.go` line 7899 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/159162#-OgEHF_R44Q3Arz46Q7j:-OgEHF_S-LHiKdU7F4Jh:b72iuxl) ([raw file](https://github.com/cockroachdb/cockroach/blob/f3086f2fadf5b0e00f9fb1aaae744266bf86e1d9/pkg/sql/sem/builtins/builtins.go#L7899)):*
> ```Go
> 				// The schema change below requires us to relesae our current timestamp, otherwise
> 				// we may get stuck waiting for leases.
> 				// TODO(fqazi): We should plumb an executor from the planner instead, we can
> ```

hm, i'm not sure if we can actually do this TODO easily.


<!-- Sent from Reviewable.io -->

fqazi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/159162)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @rafiss)

___
*[`pkg/sql/catalog/descs/leased_descriptors.go` line 202 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/159162#-OgEHSl7612D_HIj4ypt:-OgHc2jn6VGidPRvSI-f:b-yrngc4) ([raw file](https://github.com/cockroachdb/cockroach/blob/f3086f2fadf5b0e00f9fb1aaae744266bf86e1d9/pkg/sql/catalog/descs/leased_descriptors.go#L202)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: should we get a tracking issue for moving timestamps forward? it feels a bit more complex than just a TODO.
</blockquote></details>

Let me get a tracking issue open for this.
___
*[`pkg/sql/sem/builtins/builtins.go` line 7899 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/159162#-OgEHF_R44Q3Arz46Q7j:-OgHhFZeEcgKkc7mWZ7V:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f3086f2fadf5b0e00f9fb1aaae744266bf86e1d9/pkg/sql/sem/builtins/builtins.go#L7899)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

hm, i'm not sure if we can actually do this TODO easily.
</blockquote></details>

Done.
___
*[`pkg/sql/sem/eval/deps.go` line 292 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/159162#-OgEGwYP6mSPeCsZE9za:-OgHcYbvEaZl2M5IDiXy:b73y75m) ([raw file](https://github.com/cockroachdb/cockroach/blob/f3086f2fadf5b0e00f9fb1aaae744266bf86e1d9/pkg/sql/sem/eval/deps.go#L292)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

just to clarify, what is "achieve this" referring to?
</blockquote></details>

Done.

The intention was to indicate that we should just use the same transaction, otherwise there is a risk of leases being held.

//The correct way  
// to achieve this is to execute the schema change on the same transaction,  
// which also reduces the risk of waiting for leases to expire.
___
*[`pkg/sql/sem/builtins/pg_builtins.go` line 1432 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/159162#-OgEH1O87IZBHp29jRar:-OgHhG9z0XoCJL8aa2eE:b8avtan) ([raw file](https://github.com/cockroachdb/cockroach/blob/f3086f2fadf5b0e00f9fb1aaae744266bf86e1d9/pkg/sql/sem/builtins/pg_builtins.go#L1432)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

what's the impact of this TODO? meaning, if you try to do introspection after the pg_sleep, it might have weird results?
</blockquote></details>

Done.

I added this for a test, but I'll move it over to the new setting.


<!-- Sent from Reviewable.io -->

rafiss reviewed (approved):
lgtm!

> @rafiss reviewed 4 of 9 files at r6, 2 of 5 files at r7, 8 of 8 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/159162#-:-OgIVwjM1QW9WJCjVgSw:b-7nbn20)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @fqazi)



<!-- Sent from Reviewable.io -->

fqazi committed:
catalog: ensure descriptors from KV match leased descriptors

Previously, when leased descriptors were used for catalog views in the
same transaction as schema changes, it was possible for applications to
make incorrect decisions. One example is the schema changer workload
which will do catalog queries, that will influence the schema changes
allowed. If a leased descriptor has stale data then it can attempt the
same schema change or invalid schema changes based on incorrect
information. This patch adds logic to force KV descriptor reads to
ensure that versions align with leased descriptors, otherwise forcing a
retry.

Release note: None

rafiss reviewed (changes requested):
i had a few small nits on the new commit, and also some clarifying questions to make sure i understand the goal

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/159162#-:-OgXPH_v9Hmu2IPcd_7D:b-rdiww0)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @fqazi and @kyle-a-wong)

___
*[`pkg/sql/table.go` line 310 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/159162#-OgXIm-K9VYlNRnx-9ZC:-OgXIm-K9VYlNRnx-9ZD:b-g0ha9t) ([raw file](https://github.com/cockroachdb/cockroach/blob/e71472c0685ed6f168acbd6ff938c661c23efbee/pkg/sql/table.go#L310)):*
> ```Go
> 			// membership cache. ClientVisibleRetryError's do not bump the txns epochs
> 			// automatically.
> 			return p.Txn().GenerateForcedRetryableErr(ctx, redact.Sprintf("%s", err))
> ```

nit: use `%v` instead of `%s`
___
*[`pkg/sql/rolemembershipcache/cache.go` line 144 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/159162#-OgXKnhUEq0-JF7Uonnj:-OgXKnhUEq0-JF7Uonnk:b-7nr2b0) ([raw file](https://github.com/cockroachdb/cockroach/blob/e71472c0685ed6f168acbd6ff938c661c23efbee/pkg/sql/rolemembershipcache/cache.go#L144)):*
> ```Go
> ) (_ map[username.SQLUsername]bool, retErr error) {
> 	defer func() {
> 		// Convert retry on modified descriptor errors into forced retryable errors,
> ```

do you know why only the role membership cache needed this? why not other caches?
___
*[`pkg/sql/rolemembershipcache/cache.go` line 148 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/159162#-OgXMfxV3UCYiu8jiVU6:-OgXMfxV3UCYiu8jiVU7:b-2dvsjr) ([raw file](https://github.com/cockroachdb/cockroach/blob/e71472c0685ed6f168acbd6ff938c661c23efbee/pkg/sql/rolemembershipcache/cache.go#L148)):*
> ```Go
> 		// do not bump the txns epochs automatically.
> 		if descs.IsRetryOnModifiedDescriptorError(retErr) {
> 			retErr = txn.KV().GenerateForcedRetryableErr(ctx, redact.Sprintf("%s", retErr))
> ```

nit: for safety, move this `defer` so it's after the `txn == nil` check.
___
*[`pkg/sql/catalog/descs/leased_descriptors.go` line 115 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/159162#-OgXN5q_E-Gc7Gb0Soqb:-OgXN5q_E-Gc7Gb0Soqc:b-mq06pu) ([raw file](https://github.com/cockroachdb/cockroach/blob/e71472c0685ed6f168acbd6ff938c661c23efbee/pkg/sql/catalog/descs/leased_descriptors.go#L115)):*
> ```Go
> func (r *retryOnModifiedDescriptor) SafeFormatError(p errors.Printer) (next error) {
> 	p.Printf("the descriptor %s(%d) has been modified at timestamp %s, which is no longer usable by the transaction's timestamp: %s",
> 		r.descName, r.descID, r.expiration.String(), r.readTimestamp)
> ```

nit: avoid the call to `expiration.String()`. that would cause unwanted redaction. we can just use `r.expiration` here
___
*[`pkg/sql/catalog/descs/leased_descriptors.go` line 120 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/159162#-OgXKvFrAfPLtwyt5nvl:-OgXKvFrAfPLtwyt5nvm:bxlrxey) ([raw file](https://github.com/cockroachdb/cockroach/blob/e71472c0685ed6f168acbd6ff938c661c23efbee/pkg/sql/catalog/descs/leased_descriptors.go#L120)):*
> ```Go
> 
> // IsRetryOnModifiedDescriptorError checks if an error is of type retryOnModifiedDescriptor.
> func IsRetryOnModifiedDescriptorError(err error) bool {
> ```

instead of just checking the error type, could this function actually do the conversion to a ForcedRetryableError? like this:

```
func HandleRetryOnModifiedDescriptorError(ctx context.Context, txn *kv.Txn, err error) error {
  if errors.HasType(err, (*retryOnModifiedDescriptor)(nil)) {
    return txn.KV().GenerateForcedRetryableErr(ctx, redact.Sprintf("%v", err))
  }
  return err
}

```

(this comment is only relevant if we don't modify `ensureLeasedAndKVVersionsMatch` to return the ForcedRetryableError directly.)
___
*[`pkg/sql/catalog/descs/descriptor.go` line 254 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/159162#-OgXJug0EzyMk52bt_xu:-OgXJug0EzyMk52bt_xv:b-u7akqk) ([raw file](https://github.com/cockroachdb/cockroach/blob/e71472c0685ed6f168acbd6ff938c661c23efbee/pkg/sql/catalog/descs/descriptor.go#L254)):*
> ```Go
> 
> // ensureLeasedAndKVVersionsMatch ensures that a KV and leased descriptors/
> // in a given transaction have compatible versions. If they don't, then retry
> ```

i think an extra detail in the comments could be useful: if I read this correctly, it looks like this function only compares the descriptor if we already had one cached, meaning this doesn't actually do another lookup from the store. (initially I thought this was going to fetch from disk.)

but also I want to make sure I understand -- doesn't this now mean that the pg_catalog/information_schema tables will never actually have any stale data? if we're going to make the client retry anyway, doesn't that mean we lose the benefits of populating the virtual tables with leased descriptors and avoiding the lookup?

out of curiosity, what was the error that this fixes?
___
*[`pkg/sql/catalog/descs/descriptor.go` line 284 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/159162#-OgXNbqu-iUL37mqE9rl:-OgXNbqu-iUL37mqE9rm:b-jvvp19) ([raw file](https://github.com/cockroachdb/cockroach/blob/e71472c0685ed6f168acbd6ff938c661c23efbee/pkg/sql/catalog/descs/descriptor.go#L284)):*
> ```Go
> 		modificationTime = otherDescriptor.GetModificationTime()
> 	}
> 	return &retryOnModifiedDescriptor{
> ```

nit: could we just call `txn.KV().GenerateForcedRetryableErr` at this point, rather than needing the logic to convert it later?


<!-- Sent from Reviewable.io -->

fqazi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/159162)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @kyle-a-wong and @rafiss)

___
*[`pkg/sql/catalog/descs/descriptor.go` line 254 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/159162#-OgXJug0EzyMk52bt_xu:-OgXPy89ELBPDpzyVi37:b-yc7hyz) ([raw file](https://github.com/cockroachdb/cockroach/blob/e71472c0685ed6f168acbd6ff938c661c23efbee/pkg/sql/catalog/descs/descriptor.go#L254)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

i think an extra detail in the comments could be useful: if I read this correctly, it looks like this function only compares the descriptor if we already had one cached, meaning this doesn't actually do another lookup from the store. (initially I thought this was going to fetch from disk.)

but also I want to make sure I understand -- doesn't this now mean that the pg_catalog/information_schema tables will never actually have any stale data? if we're going to make the client retry anyway, doesn't that mean we lose the benefits of populating the virtual tables with leased descriptors and avoiding the lookup?

out of curiosity, what was the error that this fixes?
</blockquote></details>

This fixes the case where a crdb_internal/pg_catalog is used before a schema change is made. It doesn't force us into storage, but it guarantees if we read a descriptor via KV laterthe version should match the leased descriptor (if one is checked out). This prevents inconsistencies.  In majority of cases the retry errors will never be generated outside of workloads that are heavy on schema changes (i.e. the leased descriptor will very likely just be up to date).

The simplest example in the schema changer workload is:
 We run a select on create_statements and based on that we picked a schema change. If the leased descriptors is stale, once we read the KV descriptor the schema change we are trying to apply may be invalid. So, it's a consistency bug if we operate on contradictory information.

I updated the comment to add:
```
// ensureLeasedAndKVVersionsMatch ensures that a KV and leased descriptors/
// in a given transaction have compatible versions. This impacts transactions
// that execute catalog queries followed by schema changes, where schema changes
// always require the freshest copy from the KV store. If they don't, then retry error is forced, since there is a risk of making decisions on stale data within the application.
```
___
*[`pkg/sql/catalog/descs/descriptor.go` line 284 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/159162#-OgXNbqu-iUL37mqE9rl:-OgXQyqP3CUMhhKSm2Rp:b-kikzz8) ([raw file](https://github.com/cockroachdb/cockroach/blob/e71472c0685ed6f168acbd6ff938c661c23efbee/pkg/sql/catalog/descs/descriptor.go#L284)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: could we just call `txn.KV().GenerateForcedRetryableErr` at this point, rather than needing the logic to convert it later?
</blockquote></details>

Done.

This is much simpler, great suggestion :)
___
*[`pkg/sql/catalog/descs/leased_descriptors.go` line 120 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/159162#-OgXKvFrAfPLtwyt5nvl:-OgXR9VK-fLuGV1V6-qZ:bvu0oiv) ([raw file](https://github.com/cockroachdb/cockroach/blob/e71472c0685ed6f168acbd6ff938c661c23efbee/pkg/sql/catalog/descs/leased_descriptors.go#L120)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

instead of just checking the error type, could this function actually do the conversion to a ForcedRetryableError? like this:

```
func HandleRetryOnModifiedDescriptorError(ctx context.Context, txn *kv.Txn, err error) error {
  if errors.HasType(err, (*retryOnModifiedDescriptor)(nil)) {
    return txn.KV().GenerateForcedRetryableErr(ctx, redact.Sprintf("%v", err))
  }
  return err
}

```

(this comment is only relevant if we don't modify `ensureLeasedAndKVVersionsMatch` to return the ForcedRetryableError directly.)
</blockquote></details>

Done.

I opted to modify that function to return the force retry error instead.
___
*[`pkg/sql/rolemembershipcache/cache.go` line 144 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/159162#-OgXKnhUEq0-JF7Uonnj:-OgXRDrU0FFX6EZ-qD1k:bjtmtz0) ([raw file](https://github.com/cockroachdb/cockroach/blob/e71472c0685ed6f168acbd6ff938c661c23efbee/pkg/sql/rolemembershipcache/cache.go#L144)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

do you know why only the role membership cache needed this? why not other caches?
</blockquote></details>

Done.

It had specific logic to detect an epoch bump, this can be removed since the conversion is gone. The main case was with high priority transactions.
___
*[`pkg/sql/rolemembershipcache/cache.go` line 148 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/159162#-OgXMfxV3UCYiu8jiVU6:-OgXRRc90YeRneqamLxC:b-jf6aud) ([raw file](https://github.com/cockroachdb/cockroach/blob/e71472c0685ed6f168acbd6ff938c661c23efbee/pkg/sql/rolemembershipcache/cache.go#L148)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: for safety, move this `defer` so it's after the `txn == nil` check.
</blockquote></details>

Done.

Logic is removed.


<!-- Sent from Reviewable.io -->

rafiss reviewed (approved):
nice work! i just thought of one more tiny nit around error wrapping, but lgtm otherwise

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/159162#-:-OgYsX4h6gePjbD6nJTL:bihvni1)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @fqazi and @kyle-a-wong)

___
*[`pkg/sql/catalog/descs/descriptor.go` line 293 at r24](https://reviewable.io/reviews/cockroachdb/cockroach/159162#-OgYqZXqETf3DilttHDC:-OgYqZXqETf3DilttHDD:blq14af) ([raw file](https://github.com/cockroachdb/cockroach/blob/46d9d4e8e5a2b6c573746600432085c75785457e/pkg/sql/catalog/descs/descriptor.go#L293)):*
> ```Go
> 		descName:      descriptor.GetName(),
> 		expiration:    modificationTime,
> 		readTimestamp: txn.ReadTimestamp(),
> ```

just thought of one more small change that could help make error introspection easier in the future. `retryOnModifiedDescriptor` could have one more field so that it wraps the underlying forced retry error:

```
	err := &retryOnModifiedDescriptor{
		descID:        descriptor.GetID(),
		descName:      descriptor.GetName(),
		expiration:    modificationTime,
		readTimestamp: txn.ReadTimestamp(),
		forcedErr: txn.GenerateForcedRetryableErr(ctx, "forcing txn to retry due to modified descriptor")
	}
```

this should still work, and it would also mean that we can use `errors.HasType(retryOnModifiedDescriptor)` in case we ever need to.


<!-- Sent from Reviewable.io -->

fqazi reviewed (commented):
> @fqazi reviewed 1 of 2 files at r24.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/159162)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @kyle-a-wong and @rafiss)

___
*[`pkg/sql/catalog/descs/descriptor.go` line 293 at r24](https://reviewable.io/reviews/cockroachdb/cockroach/159162#-OgYqZXqETf3DilttHDC:-OgYuuqt5SZ7rGxBzUb7:bl0om3n) ([raw file](https://github.com/cockroachdb/cockroach/blob/46d9d4e8e5a2b6c573746600432085c75785457e/pkg/sql/catalog/descs/descriptor.go#L293)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

just thought of one more small change that could help make error introspection easier in the future. `retryOnModifiedDescriptor` could have one more field so that it wraps the underlying forced retry error:

```
	err := &retryOnModifiedDescriptor{
		descID:        descriptor.GetID(),
		descName:      descriptor.GetName(),
		expiration:    modificationTime,
		readTimestamp: txn.ReadTimestamp(),
		forcedErr: txn.GenerateForcedRetryableErr(ctx, "forcing txn to retry due to modified descriptor")
	}
```

this should still work, and it would also mean that we can use `errors.HasType(retryOnModifiedDescriptor)` in case we ever need to.
</blockquote></details>

Done.

Great suggestion!


<!-- Sent from Reviewable.io -->

fqazi commented:
@rafiss TFTR

bors r+

rafiss commented on pkg/sql/catalog/descs/leased_descriptors.go:
```diff
@@ -81,15 +82,41 @@ func makeLeasedDescriptors(lm LeaseManager) leasedDescriptors {
+func (r *retryOnModifiedDescriptor) SafeFormatError(p errors.Printer) (next error) {
+	p.Printf("the descriptor %s(%d) has been modified at timestamp %s, which is no longer usable by the transaction's timestamp: %s",
+		r.descName, r.descID, r.expiration, r.readTimestamp)
+	return nil
```
i think we want `return r.forcedErr` now

fqazi commented:
bors r-

fqazi commented:
@rafiss TFTR

bors r+
